### ISSUE TITLE
twisted.internet.process raises an exception when encoding tracebacks that occur post-fork/pre-exec and include non-ASCII characters

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7878]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7878|
|Type|enhancement|
|Created|2015-04-30 08:32:02Z|
|Branch|https://github.com/twisted/twisted/tree/error-encode-7878|

<details><summary>Searchable metadata</summary>

```
trac-id__7878 7878
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__ 
branch__branches_error_encode_7878 branches/error-encode-7878
branch_author__glyph glyph
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1430382722713470 1430382722713470
changetime__1431160674551312 1431160674551312
version__None None
owner__glyph glyph

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44629]) Branching to tip-encode-7878.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

This has the following things in it:

- Doesn't double-allocate for Py2.
- Encodes to the parent's stdio encoding on Python 3.

Builders spun, please review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set status to `closed`|
|-|-|

No problem described here.  Please try again.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set status to `reopened`|
|-|-|

Replying to [exarkun](#note_3):
> No problem described here.  Please try again.

The problem is described in the subject ("might encode tracebacks in children wrong") and then in more detail in the previous comment (a potential efficiency issue for double-allocation for encoding on python 2, and failure to honor the stderr stream's existing encoding on python 3).  If the description isn't detailed enough for you, closing as "no problem described" seems unduly harsh; just ask for a clarification if you don't understand the point of the change.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph commented|
|-|-|

So, as it happens, _I_ don't understand the point of the change :-).

The first red flag for this change is that there's no accompanying test.  If there's behavior that should be different, then there ought to be a test which simulates the conditions under which that different behavior should occur; changing the internals like this with no accompanying test doesn't give me confidence that it changes the right stuff.  So please add a test case which validates the sequence of events you expect to happen, since then I can comment intelligently not just on the ticket's prose description, but whether the precise set-up of the test corresponds to that description.

I get the impression that there are some things that can happen between import time and execution of this function which can potentially mess things up.  Extrapolating from both the earlier description and what I see in the code, it seems that you're concerned with the potential replacement of the `sys.stderr` object from something that "really" represents stderr, i.e. file descriptor 2, with something that represents a logging wrapper or something similar, which will cause the encoding to be incorrect.  This change then moves the discovery of the "actual" stderr encoding much earlier, which allows for (I assume) a more faithful transcription of the traceback into the appropriate encoding for that stream.

Except, there's one problem - the traceback most likely isn't even going to that stream.

In the `try` block, we are doing three things: `settrace`, `_setupChild`, and `_execChild`.  As the (not always entirely correct) message suggests, the _usual_ failure mode here is in `execvpe` and not in child file descriptor setup.  So, most likely, by the time we're emitting this traceback, we are emitting it to the file descriptor provisioned by twisted itself for stderr, so we can decide on our own encoding.  Since we are (probably) just emitting bytes which will be consumed by the parent process anyway, those are going to be passed to `childDataReceived` and aren't necessarily going to be presented in any particular way.

In any case, we're still emitting a traceback to a byte-stream in a panic situation.  The important thing is to select an encoding which will allow for some recovery of the data (one which won't traceback while trying to format the traceback); one which definitively can encode the full unicode character set, rather than one which will definitely match whatever happened to be stderr when we started up.  To my mind, the "wrong" in "process might encode tracebacks in children wrong", is that we might blow up trying to encode a legit traceback with a non-ascii codepoint in it (say, from a line in a UTF-8 encoded Python file) just because the parent process happened to start up with a US-ASCII locale, even if it has properly initialized a unicode logging context and can log the traceback just fine.

So, hawkowl, would you agree that the issue here is "we might traceback while tracebacking" is the encoding problem to fix here? Or is there something else you had in mind?

I feel like, in the absence of a performance test, worrying about the double-encode in py2 is perhaps wasteful; this case is only hit in a pretty extreme scenario we don't need to optimize for, so dealing with that in its own right might not be worthwhile. Having less divergent py2/py3 code paths would be worthwhile though.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @hawkowl<br>@glyph set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @glyph|
|-|-|

OK this problem is really fascinating me, I think I'll take a crack at it in a new branch.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

(In [44634]) Branching to error-encode-7878.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_11"></a>|@glyph commented|
|-|-|

Removing hawkowl as an author since I didn't base this on the previous branch at all.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Builds are looking reasonably good.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_13"></a>|@hawkowl commented|
|-|-|

These tests seem to fail on the (unsupported and new) Debian 7 3.4 and 3.5 builders. https://buildbot.twistedmatrix.com/builders/debian7-py3.4/builds/1/steps/shell/logs/stdio and https://buildbot.twistedmatrix.com/builders/debian7-py3.5/builds/2/steps/shell/logs/stdio .

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph commented|
|-|-|

This is definitely a little concerning. You mentioned on IRC that these builders were in a different locale? Was that `LANG=C` or, like, no environment variables at all?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl commented|
|-|-|

Yes:

```
buildslave@8d67a05e3f18:/buildslave$ locale
LANG=
LANGUAGE=
LC_CTYPE="POSIX"
LC_NUMERIC="POSIX"
LC_TIME="POSIX"
LC_COLLATE="POSIX"
LC_MONETARY="POSIX"
LC_MESSAGES="POSIX"
LC_PAPER="POSIX"
LC_NAME="POSIX"
LC_ADDRESS="POSIX"
LC_TELEPHONE="POSIX"
LC_MEASUREMENT="POSIX"
LC_IDENTIFIC
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_16"></a>|@glyph commented|
|-|-|

What does `env | grep -e '\(LC_\|LANG\)'` look like?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_17"></a>|@hawkowl commented|
|-|-|

It has nothing, here's the full output of `env`:

```
buildslave@9702e65615a3:/buildslave/twisted$ env
HOSTNAME=9702e65615a3
SHELL=/bin/bash
TERM=xterm
USER=buildslave
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arj=01;31:*.taz=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lz=01;31:*.xz=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.jpg=01;35:*.jpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.axv=01;35:*.anx=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.axa=00;36:*.oga=00;36:*.spx=00;36:*.xspf=00;36:
MAIL=/var/mail/buildslave
PATH=/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
PWD=/buildslave/twisted
HOME=/home/buildslave
SHLVL=2
LOGNAME=buildslave
_=/usr/bin/env
OLDPWD=/buildslave
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_18"></a>|@glyph commented|
|-|-|

OK. That's what I thought. I can't replicate this on OS X, possibly because the filesystem encoding is always utf-8 on OS X regardless of environment. Pulling some dockers now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

This is a bug in Python itself, apparently introduced in version 3.5.

```
root@1c85cfc5c359:/Users/glyph/Projects/Twisted# cat unitb.py 
import sys
import traceback

def unicode_traceback():
    # Ensure that real traceback formatting has some non-ASCII in it,
    # by forcing the filename of the last frame to contain non-ASCII.
    filename = u"\N{SNOWMAN}"
    if not isinstance(filename, str):
        filename = filename.encode("utf-8")
    codeobj = compile("1/0", filename, "single")
    eval(codeobj)

try:
    unicode_traceback()
except:
    traceback.print_exc(file=sys.stderr)
root@1c85cfc5c359:/Users/glyph/Projects/Twisted# /pyenv/versions/3.5.0a4/bin/python unitb.py 
Traceback (most recent call last):
  File "unitb.py", line 14, in <module>
    unicode_traceback()
  File "unitb.py", line 11, in unicode_traceback
    eval(codeobj)
  File "\u2603", line 1, in <module>
ZeroDivisionError: division by zero

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "unitb.py", line 16, in <module>
    traceback.print_exc(file=sys.stderr)
  File "/pyenv/versions/3.5.0a4/lib/python3.5/traceback.py", line 158, in print_exc
    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)
  File "/pyenv/versions/3.5.0a4/lib/python3.5/traceback.py", line 99, in print_exception
    type(value), value, tb, limit=limit).format(chain=chain):
  File "/pyenv/versions/3.5.0a4/lib/python3.5/traceback.py", line 457, in __init__
    capture_locals=capture_locals)
  File "/pyenv/versions/3.5.0a4/lib/python3.5/traceback.py", line 342, in extract
    f.line
  File "/pyenv/versions/3.5.0a4/lib/python3.5/traceback.py", line 271, in line
    self._line = linecache.getline(self.filename, self.lineno).strip()
  File "/pyenv/versions/3.5.0a4/lib/python3.5/linecache.py", line 16, in getline
    lines = getlines(filename, module_globals)
  File "/pyenv/versions/3.5.0a4/lib/python3.5/linecache.py", line 47, in getlines
    return updatecache(filename, module_globals)
  File "/pyenv/versions/3.5.0a4/lib/python3.5/linecache.py", line 95, in updatecache
    stat = os.stat(fullname)
UnicodeEncodeError: 'ascii' codec can't encode character '\u2603' in position 0: ordinal not in range(128)
root@1c85cfc5c359:/Users/glyph/Projects/Twisted# 
```

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

Oops, spoke too soon. Behavior same in 3.4.  Apparently you just can't rely on traceback formatting if you have non-ascii filenames. Hrm. Maybe we should call into Failure and fix the bug there?

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_21"></a>|@hawkowl commented|
|-|-|

As per IRC discussions, we need to figure this out.

I'm thinking a Python 3 builder with no locales other than POSIX would be useful. I'll make up one of those, I think.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

To quote the stdlib docs for `getline`:

    “Get line lineno from file named filename. This function will never raise an exception — it will return `''` on errors”

Apparently that is not what it does.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_23"></a>|@glyph commented|
|-|-|

I've added a minor adjustment to the branch to ensure it only hits the issues in _our_ code and does not inadvertently trigger this related issue in the standard library.

From a linux py3 environment with a null locale and therefore ascii filesystem encoding:

```
root@9668abe973da:/Users/glyph/Projects/Twisted# python admin/run-python3-tests twisted.internet.test.test_process.ProcessTestsBuilder_SelectReactorTests.test_errorDuringExec
test_errorDuringExec (twisted.internet.test.test_process.ReactorBuilder.makeTestCaseClasses.<locals>.testcase)
test_errorDuringExec ... ok

----------------------------------------------------------------------
Ran 1 test in 0.012s

OK
```

There's a bug here in the stdlib where if you have modules with non-ascii names inside a custom module loader, they won't be able to format their tracebacks on systems without a full-unicode filesystem encoding.

Buildbots are working on the change but I anticipate passing all around.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_24"></a>|@hawkowl commented|
|-|-|

I like this.

Please add a comment about why utf-8 is hard coded for the benefit of people reading this code in the future, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph commented|
|-|-|

Running the builds one more time to make sure that the build-configuration bugs are addressed and we're still passing on 3.4.  Then I'll merge.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_26"></a>|@glyph commented|
|-|-|

Aside from one stuck builder, things look good.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_27"></a>|@glyph commented|
|-|-|

(In [44685]) improve the clarity of this comment, ask hawkowl asked

refs [#7878](https://github.com/twisted/twisted/issues/7878)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_28"></a>|@glyph set status to `closed`|
|-|-|

(In [44686]) Merge error-encode-7878: encode post-fork tracebacks properly

Author: glyph

Reviewer: hawkowl

Fixes: [#7878](https://github.com/twisted/twisted/issues/7878)

In the post-fork, pre-exec context, there was some overly complex code
that attempted to encode things correctly on python 3 according to
"standard" encodings, like sys.getfilesystemencoding(), but instead just
ended up failing while trying to report its failure in edge-cases
involving non-ASCII characters in tracebacks.

This simplifies the implementation, adds more test coverage, and makes
non-ASCII characters work properly.
